File: api/src/main/java/in/kyle/mcspring/scheduler/ScheduledAnnotationSupport.java
package in.kyle.mcspring.scheduler;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;

import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Profile;
import org.springframework.scheduling.Trigger;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
import org.springframework.stereotype.Component;
import org.springframework.util.concurrent.ListenableFuture;

import java.util.Date;
import java.util.concurrent.Callable;
import java.util.concurrent.Future;
import java.util.concurrent.ScheduledFuture;

@Lazy
@Component
@AllArgsConstructor
class ScheduledAnnotationSupport extends ThreadPoolTaskScheduler {

    private final SchedulerService scheduler;

    @Override
    public ScheduledFuture<?> schedule(Runnable task, Trigger trigger) {
        return super.schedule(wrapSync(task), trigger);
    }

    @Override
    public ScheduledFuture<?> schedule(Runnable task, Date startTime) {
        return super.schedule(wrapSync(task), startTime);
    }

    @Override
    public ScheduledFuture<?> scheduleAtFixedRate(Runnable task, Date startTime, long period) {
        return super.scheduleAtFixedRate(wrapSync(task), startTime, period);
    }

    @Override
    public ScheduledFuture<?> scheduleAtFixedRate(Runnable task, long period) {
        return super.scheduleAtFixedRate(wrapSync(task), period);
    }

    @Override
    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable task, Date startTime, long delay) {
        return super.scheduleWithFixedDelay(wrapSync(task), startTime, delay);
    }

    @Override
    public ScheduledFuture<?> scheduleWithFixedDelay(Runnable task, long delay) {
        return super.scheduleWithFixedDelay(wrapSync(task), delay);
    }

    @Override
    public void execute(Runnable task) {
        super.execute(task);
    }

    @Override
    public Future<?> submit(Runnable task) {
        return super.submit(task);
    }

    @Override
    public <T> Future<T> submit(Callable<T> task) {
        return super.submit(task);
    }

    @Override
    public ListenableFuture<?> submitListenable(Runnable task) {
        return super.submitListenable(task);
    }

    @Override
    public <T> ListenableFuture<T> submitListenable(Callable<T> task) {
        return super.submitListenable(task);
    }
    
    private Runnable wrapSync(Runnable task) {
        return new WrappedRunnable(scheduler, task);
    }
    
    @AllArgsConstructor
    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
    static class WrappedRunnable implements Runnable {
        
        private final SchedulerService scheduler;
        
        @EqualsAndHashCode.Include
        private final Runnable runnable;
    
        @Override
        public void run() {
            scheduler.syncTask(runnable);
        }
    }
}

File: mcspring-api/mcspring-base/src/main/kotlin/in/kyle/mcspring/scheduler/ScheduledAnnotationSupport.kt
package `in`.kyle.mcspring.scheduler

import org.bukkit.plugin.Plugin
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean
import org.springframework.context.annotation.Lazy
import org.springframework.scheduling.Trigger
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler
import org.springframework.stereotype.Component
import org.springframework.util.concurrent.ListenableFuture
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.Future
import java.util.concurrent.ScheduledFuture

@Lazy
@Component
@ConditionalOnBean(Plugin::class)
internal class ScheduledAnnotationSupport(
        private val scheduler: SchedulerService
) : ThreadPoolTaskScheduler() {

    override fun schedule(task: Runnable, trigger: Trigger): ScheduledFuture<*> {
        @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
        return super.schedule(wrapSync(task), trigger)
    }

    override fun schedule(task: Runnable, startTime: Date): ScheduledFuture<*> {
        return super.schedule(wrapSync(task), startTime)
    }

    override fun scheduleAtFixedRate(task: Runnable, startTime: Date, period: Long): ScheduledFuture<*> {
        return super.scheduleAtFixedRate(wrapSync(task), startTime, period)
    }

    override fun scheduleAtFixedRate(task: Runnable, period: Long): ScheduledFuture<*> {
        return super.scheduleAtFixedRate(wrapSync(task), period)
    }

    override fun scheduleWithFixedDelay(task: Runnable, startTime: Date, delay: Long): ScheduledFuture<*> {
        return super.scheduleWithFixedDelay(wrapSync(task), startTime, delay)
    }

    override fun scheduleWithFixedDelay(task: Runnable, delay: Long): ScheduledFuture<*> {
        return super.scheduleWithFixedDelay(wrapSync(task), delay)
    }

    private fun wrapSync(task: Runnable): Runnable {
        return WrappedRunnable(scheduler, task)
    }

    data class WrappedRunnable(val scheduler: SchedulerService, val runnable: Runnable) : Runnable {
        override fun run() {
            scheduler.syncTask(runnable)
        }
    }
}
